source/globalCommands.py

```diff
@@ -972,8 +972,21 @@ def script_review_activate(self,gesture):
 	script_review_activate.__doc__=_("Performs the default action on the current navigator object (example: presses it if it is a button).")
 	script_review_activate.category=SCRCAT_OBJECTNAVIGATION
 
+	def _moveWithBoundsChecking(self, info, unit, direction, endPoint=None):
+		"""
+		Nondestructively moves a textInfo and returns a (newInfo, res) tuple.
+		Res is 0 if the move would be out of bounds.
+		"""
+		newInfo = info.copy()
+		res = newInfo.move(unit, direction, endPoint=endPoint)
+		if newInfo.obj.reviewBounded and newInfo.isOutOfBounds():
+			return (info, 0)
+		return (newInfo, res)
+
 	def script_review_top(self,gesture):
-		info=api.getReviewPosition().obj.makeTextInfo(textInfos.POSITION_FIRST)
+		obj = api.getReviewPosition().obj
+		pos = textInfos.POSITION_FIRSTVISIBLE if obj.reviewBounded else textInfos.POSITION_FIRST

```

Comment:
Just an idea, wonder what others think. Wouldn't this be much less complex if we have POSITION_FIRSTVISIBLE, POSITION_FIRSTABSOLUTE and POSITION_FIRST? Then in TextInfo.__init__, change POSITION_FIRST into POSITION_FIRSTABSOLUTE or POSITION_FIRSTVISIBLE, based on whether review is bounded?

Reply:
My thoughts (which might not be correct at this stage):
* This would require changing all `textInfo` objects to treat `POSITION_FIRSTABSOLUTE` as `POSITION_FIRST`. We can update the ones within core, but what about add-ons?
* This could erroneously affect other `textInfo` operations (i.e. movement) that have nothing to do with review.

Reply:
You're actually right, at least regarding the first point.